### PR TITLE
Migrate components to TypeScript

### PR DESCRIPTION
**Pull request by gsouquet**
Blocked by matrix-org/matrix-js-sdk#1853

- Migrate JumpToBottomButton to TypeScript
- Migrate ReadReceiptMarker to TypeScript
- Migrate RoomDetailList to TypeScript
- Migrate RoomUpgradeWarningBar to TypeScript
- Migrate SimpleRoomHeader to TypeScript
- Migrate TopUnreadMessagesBar to TypeScript
- Migrate AvatarSetting to TypeScript
- Migrate ChangeAvatar to TypeScript
- Migrate ChangeDisplayName to TypeScript
- Migrate DevicesPanel to TypeScript
- Migrate DevicesPanelEntry to TypeScript
- Migrate IntegrationManager to TypeScript
- Migrate ProfileSettings to TypeScript

---
This change is marked as an *internal change* (Task), so will not be included in the changelog.

Preview: https://612f466b2b6c1dac4eb5960a--matrix-react-sdk.netlify.app
⚠️ Do you trust the author of this PR? Maybe this build will steal your keys or give you malware. Exercise caution. Use test accounts.

### CODE REVIEW

File: src/components/views/rooms/ReadReceiptMarker.tsx [hunk 1]
```diff
@@ -16,61 +16,73 @@ limitations under the License.
 */
 
 import React, { createRef } from 'react';
-import PropTypes from 'prop-types';
+import { RoomMember } from 'matrix-js-sdk/src';
+
 import { _t } from '../../../languageHandler';
 import { formatDate } from '../../../DateUtils';
 import NodeAnimator from "../../../NodeAnimator";
 import * as sdk from "../../../index";
```
Comment (SimonBrandner) [inline src/components/views/rooms/ReadReceiptMarker.tsx:24]: Could we please get rid of these

File: src/components/views/rooms/RoomDetailList.tsx [hunk 2]
```diff
@@ -14,24 +14,23 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
+import React from 'react';
+import { Room } from 'matrix-js-sdk/src';
+import classNames from 'classnames';
 import * as sdk from '../../../index';
```
Comment (SimonBrandner) [inline src/components/views/rooms/RoomDetailList.tsx:20]: Could we please get rid of these

File: src/components/views/rooms/RoomUpgradeWarningBar.tsx [hunk 3]
```diff
@@ -15,41 +15,39 @@ limitations under the License.
 */
 
 import React from 'react';
-import PropTypes from 'prop-types';
+import { MatrixEvent, Room, RoomState } from 'matrix-js-sdk/src';
 import * as sdk from '../../../index';
```
Comment (SimonBrandner) [inline src/components/views/rooms/RoomUpgradeWarningBar.tsx:19]: Could we please get rid of these

File: src/components/views/settings/DevicesPanel.tsx [hunk 4]
```diff
@@ -16,8 +17,8 @@ limitations under the License.
 */
 
 import React from 'react';
-import PropTypes from 'prop-types';
 import classNames from 'classnames';
+import { IMyDevice } from "matrix-js-sdk/src/client";
 
 import * as sdk from '../../../index';
```
Comment (SimonBrandner) [inline src/components/views/settings/DevicesPanel.tsx:23]: Could we please get rid of these

File: src/components/views/settings/IntegrationManager.tsx [hunk 5]
```diff
@@ -16,71 +17,73 @@ limitations under the License.
 */
 
 import React from 'react';
-import PropTypes from 'prop-types';
 import * as sdk from '../../../index';
```
Comment (SimonBrandner) [inline src/components/views/settings/IntegrationManager.tsx:20]: Could we please get rid of these

File: src/components/views/rooms/ReadReceiptMarker.tsx [hunk 6]
```diff
@@ -16,61 +16,74 @@ limitations under the License.
 */
 
 import React, { createRef } from 'react';
-import PropTypes from 'prop-types';
+import { RoomMember } from 'matrix-js-sdk/src';
```
Comment (SimonBrandner) [inline src/components/views/rooms/ReadReceiptMarker.tsx:19]:
```suggestion
import { RoomMember } from "matrix-js-sdk/src/models/room-member";
```

I do have a feeling that not doing this has caused problems before but I might be wrong

(same in other files)

File: src/components/views/settings/ChangeAvatar.tsx [hunk 7]
```diff
@@ -15,54 +16,64 @@ limitations under the License.
...
 
-    componentDidMount() {
+    public componentDidMount(): void {
         MatrixClientPeg.get().on("RoomState.events", this.onRoomStateEvents);
     }
 
     // TODO: [REACT-WARNING] Replace with appropriate lifecycle event
-    UNSAFE_componentWillReceiveProps(newProps) { // eslint-disable-line camelcase
+    // eslint-disable-next-line
+    public UNSAFE_componentWillReceiveProps(newProps): void {
```
Comment (SimonBrandner) [inline src/components/views/settings/ChangeAvatar.tsx:76]:
```suggestion
    public UNSAFE_componentWillReceiveProps(newProps: IProps): void {
```

File: src/components/views/settings/ProfileSettings.tsx [hunk 8]
```diff
@@ -19,17 +19,33 @@ import { _t } from "../../../languageHandler";
...
 import Modal from "../../../Modal";
 import ErrorDialog from "../dialogs/ErrorDialog";
 import { replaceableComponent } from "../../../utils/replaceableComponent";
 import { mediaFromMxc } from "../../../customisations/Media";
+import AccessibleButton from '../elements/AccessibleButton';
+import AvatarSetting from './AvatarSetting';
+
+interface IProps {
+
+}
```
Comment (SimonBrandner) [inline src/components/views/settings/ProfileSettings.tsx:32]: I think you can just do:

```
export default class ProfileSettings extends React.Component<{}, IState> {
```

**Review by SimonBrandner — APPROVED** [review]
Seems sane! Just some nitpicks.

Also, CI seems to not be happy. The removal of `sdk.getComponent()` in some cases can actually cause some test failures :( It might be worth checking that if the cause isn't obvious

**Comment by turt2live** [comment]
Removed team review request while this is considered blocked.

**Review by SimonBrandner — CHANGES_REQUESTED** [review]
There still seems to be a few imports that aren't using the full-path which seems to be problematic - see https://github.com/matrix-org/matrix-react-sdk/pull/6671

File: src/components/views/rooms/RoomUpgradeWarningBar.tsx [hunk 9]
```diff
@@ -15,41 +15,40 @@ limitations under the License.
 */
 
 import React from 'react';
-import PropTypes from 'prop-types';
-import * as sdk from '../../../index';
+import { MatrixEvent, Room, RoomState } from 'matrix-js-sdk/src';
```
Comment (SimonBrandner) [inline src/components/views/rooms/RoomUpgradeWarningBar.tsx:18]: `/models`?

File: src/components/views/settings/DevicesPanelEntry.tsx [hunk 10]
```diff
@@ -15,46 +16,42 @@ limitations under the License.
 */
 
 import React from 'react';
-import PropTypes from 'prop-types';
+import { IMyDevice } from 'matrix-js-sdk/src';
```
Comment (SimonBrandner) [inline src/components/views/settings/DevicesPanelEntry.tsx:19]: Here too please

File: src/components/views/settings/ChangeAvatar.tsx [hunk 11]
```diff
@@ -15,54 +16,64 @@ limitations under the License.
 */
 
 import React from 'react';
-import PropTypes from 'prop-types';
+import { MatrixEvent, Room } from 'matrix-js-sdk/src';
```
Comment (SimonBrandner) [inline src/components/views/settings/ChangeAvatar.tsx:19]: Here too, please

**Review by SimonBrandner — APPROVED** [review]
LGTM

**Review by turt2live — APPROVED** [review]
lgtm, though the copyright header stuff might change again 🙃 (internal discussion)

**Pull request merged**